File: ios/Earwyrm/Features/ArtGallery/ArtGalleryViewModel.swift
import SwiftUI

/// Single source of truth for art gallery state per lyric.
/// Used by ShareModalView, EditLyricView, and ArtworkSectionView.
@Observable
@MainActor
final class ArtGalleryViewModel {
    enum CardStyle: Equatable {
        case none
        case coverArt
        case aiVariant(Int)
    }

    var variants: [(variant: CardArtService.Variant, image: UIImage)] = []
    var selectedStyle: CardStyle = .none
    private(set) var isGeneratingArt = false
    var aiArtError: String?
    var artRemaining: Int?
    var needsUpgrade = false
    var wasFreeTierGen = false

    // Cover art state
    var coverArtImage: UIImage?
    var coverArtUrl: String?

    /// The currently active image based on selectedStyle.
    var activeImage: UIImage? {
        switch selectedStyle {
        case .none:
            return nil
        case .coverArt:
            return coverArtImage
        case .aiVariant(let index):
            guard index < variants.count else { return nil }
            return variants[index].image
        }
    }

    /// The URL of the currently active art.
    var activeUrl: String? {
        switch selectedStyle {
        case .none:
            return nil
        case .coverArt:
            return coverArtUrl
        case .aiVariant(let index):
            guard index < variants.count else { return nil }
            return variants[index].variant.imageUrl
        }
    }

    /// Backward-compatible: delegates to activeImage.
    var aiArtImage: UIImage? { activeImage }

    /// The URL of the currently selected AI variant (backward compat).
    var activeVariantUrl: String? { activeUrl }

    /// Index into AI variants array (backward compat for ArtGalleryStrip).
    var activeVariantIndex: Int {
        get {
            if case .aiVariant(let i) = selectedStyle { return i }
            return -1
        }
        set {
            selectedStyle = .aiVariant(newValue)
        }
    }

    var hasAIArt: Bool {
        !variants.isEmpty
    }

    var hasAnyArt: Bool {
        hasAIArt || coverArtImage != nil
    }

    /// Fetch all variants for a lyric and download images concurrently.
    func loadVariants(for lyric: Lyric) async {
        coverArtUrl = lyric.coverArtUrl

        // Download cover art and AI variants concurrently
        async let coverArtTask: UIImage? = {
            guard let urlStr = lyric.coverArtUrl, let url = URL(string: urlStr) else { return nil }
            return await CardArtService.downloadImage(from: url)
        }()

        async let variantsTask: [(variant: CardArtService.Variant, image: UIImage)] = {
            let fetchedVariants = await CardArtService.fetchVariants(lyricId: lyric.id)
            guard !fetchedVariants.isEmpty else { return [] }

            var loaded: [(variant: CardArtService.Variant, image: UIImage)] = []
            await withTaskGroup(of: (Int, CardArtService.Variant, UIImage?).self) { group in
                for (index, variant) in fetchedVariants.enumerated() {
                    group.addTask {
                        guard let url = URL(string: variant.imageUrl) else { return (index, variant, nil) }
                        let image = await CardArtService.downloadImage(from: url)
                        return (index, variant, image)
                    }
                }
                var results: [(Int, CardArtService.Variant, UIImage?)] = []
                for await result in group {
                    results.append(result)
                }
                results.sort { $0.0 < $1.0 }
                loaded = results.compactMap { (_, variant, image) in
                    guard let image else { return nil }
                    return (variant: variant, image: image)
                }
            }
            return loaded
        }()

        let (downloadedCoverArt, loadedVariants) = await (coverArtTask, variantsTask)

        coverArtImage = downloadedCoverArt
        variants = loadedVariants

        // Determine initial selectedStyle from lyric.cardArtUrl
        if let cardArtUrl = lyric.cardArtUrl {
            if cardArtUrl == lyric.coverArtUrl {
                selectedStyle = .coverArt
            } else if let aiIndex = variants.firstIndex(where: { $0.variant.imageUrl == cardArtUrl }) {
                selectedStyle = .aiVariant(aiIndex)
            } else if !variants.isEmpty {
                selectedStyle = .aiVariant(0)
            } else if coverArtImage != nil {
                selectedStyle = .coverArt
            } else {
                selectedStyle = .none
            }
        } else {
            selectedStyle = .none
        }
    }

    /// Generate new art, prepend to variants, select it.
    func generate(lyric: Lyric, note: String?, refinement: String? = nil) async {
        isGeneratingArt = true
        aiArtError = nil
        needsUpgrade = false
        wasFreeTierGen = false

        do {
            let result = try await CardArtService.generateArt(lyric: lyric, note: note, refinement: refinement)
            artRemaining = result.remaining
            wasFreeTierGen = result.isFreeTier
            let image = await CardArtService.downloadImage(from: result.url)

            if let image {
                let newVariant = CardArtService.Variant(
                    id: UUID(),
                    imageUrl: result.url.absoluteString,
                    createdAt: Date()
                )
                variants.insert((variant: newVariant, image: image), at: 0)
                selectedStyle = .aiVariant(0)
            }

            Analytics.track(.aiArtGenerated)
        } catch CardArtService.CardArtError.upgradeRequired {
            needsUpgrade = true
        } catch {
            aiArtError = error.localizedDescription
            print("AI art generation failed: \(error)")
        }

        isGeneratingArt = false
    }

    /// Persist the active art selection.
    func persistActiveVariant(lyricId: UUID) async {
        switch selectedStyle {
        case .none:
            await CardArtService.clearActiveVariant(lyricId: lyricId)
        case .coverArt:
            if let url = coverArtUrl {
                await CardArtService.setActiveVariant(lyricId: lyricId, imageUrl: url)
            }
        case .aiVariant(let index):
            guard index < variants.count else { return }
            let url = variants[index].variant.imageUrl
            await CardArtService.setActiveVariant(lyricId: lyricId, imageUrl: url)
        }
    }

    /// Determine what action to take based on subscription status and existing art.
    enum ArtAction {
        case showGenSheet
        case showPaywall
    }

    func resolveAction(isPlus: Bool, freeGenExhausted: Bool) -> ArtAction {
        // Free user who already used their gen or has existing art → paywall
        if !isPlus && (freeGenExhausted || wasFreeTierGen || hasAIArt) {
            return .showPaywall
        }
        // Everyone else gets the generation sheet (note entry + optional art direction)
        return .showGenSheet
    }
}

File: android/app/src/main/java/com/earwyrm/app/feature/artgallery/ArtGalleryViewModel.kt
package com.earwyrm.app.feature.artgallery

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.earwyrm.app.core.model.Lyric
import com.earwyrm.app.core.network.ArtVariant
import com.earwyrm.app.core.network.CardArtService
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import javax.inject.Inject

enum class CardStyle {
    NONE,
    AI_VARIANT
}

data class ArtGalleryState(
    val variants: List<ArtVariant> = emptyList(),
    val selectedStyle: CardStyle = CardStyle.NONE,
    val selectedVariantIndex: Int = -1,
    val isGenerating: Boolean = false,
    val error: String? = null,
    val artRemaining: Int? = null,
    val needsUpgrade: Boolean = false,
    val wasFreeTierGen: Boolean = false
)

@HiltViewModel
class ArtGalleryViewModel @Inject constructor(
    private val cardArtService: CardArtService
) : ViewModel() {

    private val _state = MutableStateFlow(ArtGalleryState())
    val state: StateFlow<ArtGalleryState> = _state.asStateFlow()

    val hasAIArt: Boolean get() = _state.value.variants.isNotEmpty()

    val activeVariantUrl: String?
        get() {
            val s = _state.value
            return if (s.selectedStyle == CardStyle.AI_VARIANT && s.selectedVariantIndex in s.variants.indices) {
                s.variants[s.selectedVariantIndex].imageUrl
            } else null
        }

    fun loadVariants(lyric: Lyric) {
        viewModelScope.launch {
            val variants = cardArtService.fetchVariants(lyric.id)
            val selectedIndex: Int
            val selectedStyle: CardStyle

            if (lyric.cardArtUrl != null) {
                val aiIndex = variants.indexOfFirst { it.imageUrl == lyric.cardArtUrl }
                if (aiIndex >= 0) {
                    selectedIndex = aiIndex
                    selectedStyle = CardStyle.AI_VARIANT
                } else if (variants.isNotEmpty()) {
                    selectedIndex = 0
                    selectedStyle = CardStyle.AI_VARIANT
                } else {
                    selectedIndex = -1
                    selectedStyle = CardStyle.NONE
                }
            } else {
                selectedIndex = -1
                selectedStyle = CardStyle.NONE
            }

            _state.value = _state.value.copy(
                variants = variants,
                selectedStyle = selectedStyle,
                selectedVariantIndex = selectedIndex
            )
        }
    }

    fun selectVariant(index: Int) {
        _state.value = _state.value.copy(
            selectedStyle = CardStyle.AI_VARIANT,
            selectedVariantIndex = index
        )
    }

    fun selectNone() {
        _state.value = _state.value.copy(
            selectedStyle = CardStyle.NONE,
            selectedVariantIndex = -1
        )
    }

    fun generate(lyric: Lyric, note: String? = null, refinement: String? = null) {
        viewModelScope.launch {
            _state.value = _state.value.copy(
                isGenerating = true,
                error = null,
                needsUpgrade = false
            )

            when (val result = cardArtService.generateArt(lyric, note, refinement)) {
                is CardArtService.GenerateResult.Success -> {
                    // Refetch variants to get the new one
                    val updatedVariants = cardArtService.fetchVariants(lyric.id)
                    _state.value = _state.value.copy(
                        variants = updatedVariants,
                        selectedStyle = CardStyle.AI_VARIANT,
                        selectedVariantIndex = 0,
                        isGenerating = false,
                        artRemaining = result.remaining,
                        wasFreeTierGen = result.isFreeTier
                    )
                }
                is CardArtService.GenerateResult.UpgradeRequired -> {
                    _state.value = _state.value.copy(
                        isGenerating = false,
                        needsUpgrade = true
                    )
                }
                is CardArtService.GenerateResult.Error -> {
                    _state.value = _state.value.copy(
                        isGenerating = false,
                        error = result.message
                    )
                }
            }
        }
    }

    fun persistSelection(lyricId: String) {
        viewModelScope.launch {
            val s = _state.value
            if (s.selectedStyle == CardStyle.NONE) {
                cardArtService.clearActiveVariant(lyricId)
            } else if (s.selectedStyle == CardStyle.AI_VARIANT && s.selectedVariantIndex in s.variants.indices) {
                cardArtService.setActiveVariant(lyricId, s.variants[s.selectedVariantIndex].imageUrl)
            }
        }
    }

    fun clearUpgradeFlag() {
        _state.value = _state.value.copy(needsUpgrade = false)
    }

    fun resolveAction(isPlus: Boolean, freeGenExhausted: Boolean): ArtAction {
        if (!isPlus && (freeGenExhausted || _state.value.wasFreeTierGen || hasAIArt)) {
            return ArtAction.SHOW_PAYWALL
        }
        return ArtAction.SHOW_GEN_SHEET
    }

    enum class ArtAction {
        SHOW_GEN_SHEET,
        SHOW_PAYWALL
    }
}
